docs(tutorial): awsim instructions (#169)
autowarefoundation/autoware-documentation: docs/tutorials/digital-twin-simulation/awsim-tutorial.md

# AWSIM simulator setup

## Preparation

- Install Autoware following the [official guide](https://autowarefoundation.github.io/autoware-documentation/main/installation/autoware/source-installation/)
- Download and extract the [AWSIM binary](https://drive.google.com/file/d/1jgB_A5cSso9dzcNffQYthe07GG07QLQh/view?usp=sharing)
- Download and extract the [sample map folder](https://drive.google.com/file/d/1vGFI0o0zQ-gRZYqKrPbnrtCN3c3-92Fy/view?usp=sharing)

## Running steps

### AWSIM Simulator

> To run the AWSIM simulator, open a terminal window, navigate to the simulation directory and run:

Reviewer: ```suggestion
To run the AWSIM simulator, open a terminal window, navigate to the simulation directory and run the following commands:
```

> To run the AWSIM simulator, open a terminal window, navigate to the simulation directory and run:

Reviewer: ```suggestion
To run the AWSIM simulator, open a terminal window, navigate to the simulation directory and run the following commands:
```

```bash
source /opt/ros/galactic/setup.bash
cd <PATH_TO_EXTRACTED_FOLDER_WITH_AWSIM>
./AWSIM.x86_64
```

If the simulator is working properly, Lidar rays should be visible and the [expected data topics](https://github.com/autowarefoundation/autoware-projects/wiki/Bus-ODD-Simulation-requirements#integration-interface-with-autoware-universe) should appear when `ros2 topic list` is run.

![correctly working simulator](./images/workingSim.png)

#### Simulator shortcuts

> It is possible to adjust/control the simulation through the shortcuts listed below:

Reviewer: ```suggestion
It is also possible to control the simulation using the keyboard shortcuts listed below:
```

> | Shortcut | Description                                                |
> | -------- | ---------------------------------------------------------- |
> | v        | turn on/off lidar points visualization                     |
> | m        | enter/exit manual operation mode (Ego listens to keyboard) |
> |          | <b> In manual mode: </b>                                   |
> | d        | drive                                                      |
> | p        | parking brake                                              |
> | r        | reverse                                                    |
> | arrows   | acceleration and steering                                  |
> | 1        | left blinker                                               |
> | 2        | right blinker                                              |
> | 3        | emergency lights                                           |
> | 4        | turning the lights off                                     |

Reviewer: - Suggestion to separate out the commands from non-manual mode and manual mode
- Do the light controls (left/right blinker, emergency lights) act as toggles? So, pressing `1` turns on the left blinker and it will stay on until `1` is pressed again? Or does `4` turn the blinker off?

```suggestion
| Shortcut | Description                                                |
| -------- | ---------------------------------------------------------- |
| v        | Toggle lidar points visualization                          |
| m        | Toggle manual operation mode (Ego listens to keyboard)     |

<b> In manual mode: </b>
| Shortcut | Description                                                |
| -------- | ---------------------------------------------------------- |
| d        | Shift gear to Drive                                                      |
| p        | Shift gear to Park                                                       |
| r        | Shift gear to Reverse                                                    |
| ← ↑ ↓ → | Accelerate/brake and steering                                  |
| 1        | Left blinker                                               |
| 2        | Right blinker                                              |
| 3        | Emergency lights                                           |
| 4        | Turn off lights                                            |
```

![how to initialize autoware](./images/initializeAutoware.png)

- To make Autoware plan the route, use the `2D Goal Pose` button to set a goal pose somewhere within the Lanelet2 road lanes.

- To make Autoware execute the planned route, open a new terminal and send the engage message:

```bash
cd <YOUR_WORKSPACE>/autoware
source install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true"
```

From now on, Autoware will execute the planned route and operate the AWSIM Simulator.

![AWSIM with autoware](./images/awf_awsim.png)

### NPC Control (optional)

Currently the only supported way to add NPCs to the simulation is using prerecorded rosbags.

#### Preparation

- Download and extract [npc_msgs](https://drive.google.com/file/d/1fMvjOnz7Z0cGXotwdOAhBfLPYO52ssp0/view?usp=sharing)
- Download and extract [sample rosbags](https://drive.google.com/file/d/1BSoIq82DW_RIyeSR6OinhKmA6lD6jmWe/view?usp=sharing)

#### Run NPCs from rosbag

To run NPCs from rosbag, follow the steps:

> - navigate to the directory with extracted npc_msgs
> - build and source npc_msgs:

Reviewer: ```suggestion
- Navigate to the directory where the `npc_msgs` source code was extracted
- Build and source `npc_msgs`:
```

```bash
colcon build
source install/setup.bash
```

> - play sample rosbag while AWSIM is running:

Reviewer: ```suggestion
- Play the downloaded sample rosbag while AWSIM is running:
```

```bash
ros2 bag play <ROSBAG_NAME>
```

![AWSIM with autoware and NPCs](./images/awsim_npc.png)

### Troubleshooting

> If the data on topics is not produced or the poincloud is not visible on the simulation screen please report a problem in autoware [Q&A discussion](https://github.com/autowarefoundation/autoware/discussions/). To make analysis of the problem easier, please share a `Player.log` file (the file can be found under `~/.config/unity3d/Tier\ IV/E2ESimulator/Player.log` directory).

Reviewer 2: The following must be fixed from spell check:
`poincloud` -> `pointcloud`

Reviewer: was curious about which one is correct - from what I can see, `pointcloud` is used to refer to the message type (more specifically, `PointCloud2`) but otherwise it should be two words - `point cloud`.

Reviewer 2: Good point.
It looks like ROS PCL wiki uses "point cloud" so let's stick with two words in the text.

> If the data on topics is not produced or the poincloud is not visible on the simulation screen please report a problem in autoware [Q&A discussion](https://github.com/autowarefoundation/autoware/discussions/). To make analysis of the problem easier, please share a `Player.log` file (the file can be found under `~/.config/unity3d/Tier\ IV/E2ESimulator/Player.log` directory).

Reviewer: Added some more detail about what should be added to a problem report (env details, repro steps, screenshots etc), to avoid any issues where we just get a statement along the lines of "Point cloud is not displayed" and then a Player.log file.

```suggestion
If the expected ROS topic data is not produced, or the point cloud is not visible on the simulation screen, report the problem as an  [Autoware Q&A discussion](https://github.com/autowarefoundation/autoware/discussions/). 

To make analysis of the problem easier, make sure to provide a description of your environment, a detailed set of reproduction steps, screenshots or a short video of any unexpected behaviour and attach the `Player.log` file, which can be found in the `~/.config/unity3d/Tier\ IV/E2ESimulator/Player.log` directory.
```

```

From now on, Autoware will execute the planned route and operate the AWSIM Simulator.

![AWSIM with autoware](./images/awf_awsim.png)

### NPC Control (optional)

Currently the only supported way to add NPCs to the simulation is using prerecorded rosbags.

#### Preparation

- Download and extract [npc_msgs](https://drive.google.com/file/d/1fMvjOnz7Z0cGXotwdOAhBfLPYO52ssp0/view?usp=sharing)
- Download and extract [sample rosbags](https://drive.google.com/file/d/1BSoIq82DW_RIyeSR6OinhKmA6lD6jmWe/view?usp=sharing)

#### Run NPCs from rosbag

To run NPCs from rosbag, follow the steps:

- Navigate to the directory where the `npc_msgs` source code was extracted
- Build and source `npc_msgs`:

```bash
colcon build
source install/setup.bash
```

- Play the downloaded sample rosbag while AWSIM is running:

```bash
ros2 bag play <ROSBAG_NAME>
```

![AWSIM with autoware and NPCs](./images/awsim_npc.png)

### Troubleshooting

If the expected ROS topic data is not produced, or the point cloud is not visible on the simulation screen, report the problem as an [Autoware Q&A discussion](https://github.com/autowarefoundation/autoware/discussions/).

> To make analysis of the problem easier, make sure to provide a description of your environment, a detailed set of reproduction steps, screenshots or a short video of any unexpected behaviour and attach the `Player.log` file, which can be found in the `~/.config/unity3d/Tier\ IV/E2ESimulator/Player.log` directory.

Reviewer 3: @LalithVipulananthan FYI, TIER IV's brand guideline has changed from `Tier IV` to `TIER IV`. So, could you check it from next time on? :pray: 
https://www.tier4.jp/media/detail/?sys_id=4nwhhCgZGuFnqt9MuveSzp&category=DOCUMENTS

Reviewer: @WJaworskiRobotec Could you confirm if the `Player.log` path is correct please, specifically the `Tier\IV` part of the path?

# AWSIM Simulator Setup

This instruction presents how to run Tier IVs AWSIM with Autoware.

## Preparation

- Install Autoware following the official instruciton : [link](https://autowarefoundation.github.io/autoware-documentation/main/installation/autoware/source-installation/)
- Download and extract AWSIM binary : [link](https://drive.google.com/file/d/1jgB_A5cSso9dzcNffQYthe07GG07QLQh/view?usp=sharing)
- Download and extract map folder : [link](https://drive.google.com/file/d/1vGFI0o0zQ-gRZYqKrPbnrtCN3c3-92Fy/view?usp=sharing)

## Running steps

### AWSIM Simulator

The AWSIM Simulator can be run without any further preparation. To run the simulator:

- open the terminal navigate to the simulation directory and run:

        source /opt/ros/galactic/setup.bash
        cd <PATH_TO_EXTRACTED_FOLDER_WITH_AWSIM>
        ./AWSIM.x86_64

> The properly working simulator should have the lidar rays visible and the data topics should appear when discovering using `ros2 topic list`.

Reviewer: My suggestion says `the expected data topics` but are there any data topics that are specific to AWSIM that would indicate that is running correctly? If so, then listing those topics would be helpful to anyone running the simulator who wanted to be sure that it was working correctly.

```suggestion
If the simulator is working properly, Lidar rays should be visible and the expected data topics should appear when `ros2 topic list` is run.
```

Author: Hmm, We have created the simulation compatibility tests in other issue. It is still in progress but in general it's the package that runs sets of tests with running simulator to evaluate if interface is correct. I was planning to add the instruction about these tests in the same folder, and then maybe link the compatibility tests here ?

Author: I would keep it for now like it is, and link to in to the instruction with compatibility tests/interfaces in next PR with instructions for compatibility tests.

![correctly working simulator](./images/workingSim.png)

#### Simulator Shourtcuts

It is possible to adjust/control the simulation through the shortcuts listed below:

- v - turn on/off lidar points visualization
- m - enter/exit manual operation mode (Ego listens to keyboard)
  - In manual mode:
    - d - drive
    - p - parking brake
    - r - reverse
    - arrows - acceleration and steering
    - lights:
      - 1 - left blinker
      - 2- right blinker
      - 3 - emergency lights
      - 4- turning the lights off

> ### Autoware

Reviewer: The title of this section does not describe its purpose. Is it correct to say that this section describes how to use Autoware to control AWSIM?

> To run the Autoware, please:

Reviewer: ```suggestion
To run Autoware, open a terminal and run the following commands:
```

> - open the terminal and go to Autoware directory
> 
>         cd <YOUR_WORKSPACE>/autoware
> 
> - source the installation and run the solution
> 
>         source install/setup.bash
>         ros2 launch autoware_launch autoware.launch.xml map_path:=<PATH_TO_EXTRACTED_FOLDER_WITH_MAP> vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit use_sim_time:=true launch_sensing_driver:=false

Reviewer: ```suggestion
```console
cd <YOUR_WORKSPACE>/autoware
source install/setup.bash
ros2 launch autoware_launch autoware.launch.xml map_path:=<PATH_TO_EXTRACTED_FOLDER_WITH_MAP> vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit use_sim_time:=true launch_sensing_driver:=false
```
```

> _NOTE: The map_path has to start with /home/... , relative path will not work properly_

Reviewer: Change to use an admonition block.

```suggestion
!!! note

    Note that `map_path` must be a full path. Relative paths will not work.
```

> - the RViZ with the map should appear. The ego can localize itself at the beginning however the localization can be inaccurate. It is advised to place the `2D Pose Estimate` marker as shown on the image below.

Reviewer: ```suggestion
- RViz should appear with the Lanelet2 map displayed. Although the ego vehicle can localize itself at the beginning, the localization can be inaccurate. It is recommended to explicitly set the vehicle's initial pose using the `2D Pose Estimate` button as shown in the image below.
```

![how to initialize autoware](./images/initializeAutoware.png)

> - to make the Autoware plan the route please place the `2D Goal Pose` marker somewhere within the lanelet2 road lanes. _NOTE: The plan can be not visible due to problem with plan visualization plugin which was omitted during the Autoware build process_

Reviewer: ```suggestion
- To make Autoware plan the route, use the `2D Goal Pose` button to set a goal pose somewhere within the Lanelet2 road lanes. 

!!! note

    Note that the planned route will not be shown due to a problem with the plan visualization plugin that was omitted during the Autoware build process.
```

> - to make the Autoware execute plan please open new terminal and send the engage message:

Reviewer: ```suggestion
- To make Autoware execute the planned route, open a new terminal and send the engage message:
```

>         cd <YOUR_WORKSPACE>/autoware
>         source install/setup.bash
>         ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true"

Reviewer: ```suggestion
```console
cd <YOUR_WORKSPACE>/autoware
source install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true"
```
```

> From now on the Autoware will execute the path and operate the AWSIM Simulator.

Reviewer: ```suggestion
From now on, Autoware will execute the planned route and operate the AWSIM Simulator.
```

> ![AWSIM with autoware](./images/awf_awsim.png)

Reviewer: This image link is broken.

Author: Sorry, i forgot to upload the image. fixed now

#### Troubleshooting

> If the data on topics is not produced or the poincloud is not visible on the simulation screen please find a `Player.log` file and provide it to the simulation suppliers (the file can be found under `~/.config/unity3d/Tier\ IV/E2ESimulator` directory).

Reviewer: Instead of asking people to contact the "simulation suppliers" (TIER IV?), I think it would be better to suggest creating a Q&A discussion to report any issues and ask for assistance there.

Also, it would be better to explicitly state where the `player.log` file is within the installation folder, rather than telling people to go and look for it.

Author: Goog point. Maybe the sentence is not clear but the Player.log is a file generated by unity and it's exactly in ~/.config/unity3d/Tier\ IV/E2ESimulator/Player.log
